udkm1Dsim/simulations/xrays.py

```diff
@@ -2733,3 +2733,250 @@ def calc_roughness_matrix(roughness, k_z, last_k_z):
+
+        Returns:
+            (tuple):
+            - *chi (ndarray[complex])* - dielectric susceptibility tensor.
+            - *chi_zero (ndarray[complex])* - mean dielectric susceptibility.
+
+        """
+        try:
+            magnetization = args[0]
+            mag_amplitude = magnetization[0]
+            mag_phi = magnetization[1]
+            mag_gamma = magnetization[2]
+        except IndexError:
+            # here we catch magnetizations with only one instead of three
+            # elements
+            try:
+                mag_amplitude = atom.mag_amplitude
+            except AttributeError:
+                mag_amplitude = 0
+            try:
+                mag_phi = atom._mag_phi
+            except AttributeError:
+                mag_phi = 0
+            try:
+                mag_gamma = atom._mag_gamma
+            except AttributeError:
+                mag_gamma = 0
+
+        M = len(self._energy)  # number of energies
+        N = np.shape(self._qz)[1]  # number of q_z
+
+        chi = np.zeros([M, N, 3, 3], dtype=np.cfloat)
+        chi_zero = np.zeros([M, N], dtype=np.cfloat)
+
+        energy = self._energy
+        wl = self._wl
+
+        try:
+            cf = atom.get_atomic_form_factor(energy)
+        except AttributeError:
+            cf = np.zeros_like(energy, dtype=np.cfloat)
+        try:
+            mf = atom.get_magnetic_form_factor(energy)
+        except AttributeError:
+            mf = np.zeros_like(energy, dtype=np.cfloat)
+
+        try:
+            molar_density = density/(atom.mass_number_a/1000)*6.0222e23
+        except AttributeError:
+            molar_density = 0
+        r0 = 2.82e-15  # classical electron radius
+        multiplier = wl**2*r0/np.pi
+
+        chi_zero = molar_density*multiplier*cf
+
+        B = molar_density*multiplier*mf
+        C = 0
+        # This needs to be set in the case that a quadratic term is present
+        # And should be set in the initialization routine as a global parameter
+
+        # magnetization is possibly in cartesian coordinates
+        # need to be converted
+
+        # m1=np.array(M[0,...])
+        # m2=np.array(M[1,...])
+        # m3=np.array(M[2,...])
+        # empty=np.zeros(m1.shape)
+        # ones=empty+1
+        # delta=np.array([[ones,empty,empty],
+        #                 [empty,ones,empty],
+        #                 [empty,empty,ones]])
+        # m_epsilon1=np.array([[empty,empty,empty],
+        #         [empty,empty,m1],
+        #         [empty,-m1,empty]])
+        # m_epsilon2=np.array([[empty,empty,-m2],
+        #         [empty,empty,empty],
+        #         [m2,empty,empty]])
+        # m_epsilon3=np.array([[empty,m3,empty],
+        #         [-m3,empty,empty],
+        #         [empty,empty,empty]])
+        # temp=m_epsilon1+m_epsilon2+m_epsilon3
+        # MM = np.array([[M[j,...]*M[i,...] for i in range(3)] for j in range(3)])
+        # chi = (chi_zero)*delta\
```

Comment:
when comparing to Eq.(17) of the paper, there is a term `(chi_zero +A)` but here the A (Eq. (19)) is missing.